src/read.rs

```diff
@@ -0,0 +1,63 @@
+use std::io;
+use std::io::Write;
+use std::marker::PhantomData;
+use Event;
+
+pub struct ReadHtml<'a, I> where I: Iterator<Item=Event<'a>> + 'a {
+    events: I,
+    current: Vec<u8>,
+    phantom: PhantomData<&'a I>,
+}
+
+impl<'a, I> ReadHtml<'a, I> where I: Iterator<Item=Event<'a>> {
+    pub fn new<II>(events: II) -> ReadHtml<'a, I> where II: IntoIterator<IntoIter=I, Item=Event<'a>> {
+        ReadHtml {
+            events: events.into_iter(),
+            current: Vec::new(),
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<'a, I> io::Read for ReadHtml<'a, I> where I: Iterator<Item=Event<'a>> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        // TODO handle buf.len() == 0
```

Comment:
I think the current implementation handles it fine, from the docs for `io::Read`:

> If the return value of this method is Ok(n) [...] If n is 0, then it can indicate [...] The buffer specified was 0 bytes in length.